{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing TP53 Exon-Exon Junction Data\n",
    "\n",
    "**Gregory Way 2017**\n",
    "\n",
    "Reading in junction and sample files and merging with _TP53_ mutation data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "alt_folder = os.environ.get('PANCANCER_ALT_FOLDER')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rail_id(row):\n",
    "    \"\"\"\n",
    "    Extract specific rail_ids from complex data structure that assigns rail_ids\n",
    "    (Sample IDs) to snaptron_ids (exon-exon junctions). Designed to be used\n",
    "    as a pd.DataFrame().apply() function.\n",
    "\n",
    "    Arguments:\n",
    "    row - a row in the junction dataframe\n",
    "\n",
    "    Output: a list of sample IDs with the specific snaptron ID\n",
    "    \"\"\"\n",
    "\n",
    "    row = row['samples'].split(',')\n",
    "\n",
    "    all_sample_ids = []\n",
    "    for sample_id in row:\n",
    "        if sample_id != '':\n",
    "            sample_id = sample_id.split(':')[0]\n",
    "            all_sample_ids.append(sample_id)\n",
    "    return(all_sample_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## First, load several files required for processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rail_id</th>\n",
       "      <th>gdc_cases.samples.submitter_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TCGA-CD-8534-01A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-ER-A19A-06A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rail_id gdc_cases.samples.submitter_id\n",
       "0        0               TCGA-CD-8534-01A\n",
       "1        1               TCGA-ER-A19A-06A"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load Sample File\n",
    "sample_file = os.environ.get('PANCANCER_SNAPTRON_SAMPLES_FILE') or 'samples.tsv.gz'\n",
    "\n",
    "dictionary_df = (\n",
    "    pd.read_table(sample_file, low_memory=False)\n",
    "    .loc[:, ['rail_id', 'gdc_cases.samples.submitter_id']]\n",
    ")\n",
    "dictionary_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DataSource:Type</th>\n",
       "      <th>snaptron_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>annotated</th>\n",
       "      <th>left_motif</th>\n",
       "      <th>right_motif</th>\n",
       "      <th>left_annotated</th>\n",
       "      <th>right_annotated</th>\n",
       "      <th>samples</th>\n",
       "      <th>samples_count</th>\n",
       "      <th>coverage_sum</th>\n",
       "      <th>coverage_avg</th>\n",
       "      <th>coverage_median</th>\n",
       "      <th>source_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA:I</td>\n",
       "      <td>13874804</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7170387</td>\n",
       "      <td>7668636</td>\n",
       "      <td>498250</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>GC</td>\n",
       "      <td>AG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>,1978:1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA:I</td>\n",
       "      <td>13874806</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7170396</td>\n",
       "      <td>7668645</td>\n",
       "      <td>498250</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AT</td>\n",
       "      <td>AC</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>,11160:1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DataSource:Type  snaptron_id chromosome    start      end  length strand  \\\n",
       "0          TCGA:I     13874804      chr17  7170387  7668636  498250      -   \n",
       "1          TCGA:I     13874806      chr17  7170396  7668645  498250      +   \n",
       "\n",
       "   annotated left_motif right_motif left_annotated right_annotated   samples  \\\n",
       "0          0         GC          AG              0               0   ,1978:1   \n",
       "1          0         AT          AC              0               0  ,11160:1   \n",
       "\n",
       "   samples_count  coverage_sum  coverage_avg  coverage_median  \\\n",
       "0              1             1           1.0              1.0   \n",
       "1              1             1           1.0              1.0   \n",
       "\n",
       "   source_dataset_id  \n",
       "0                  4  \n",
       "1                  4  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load junctions file\n",
    "junction_file = os.environ.get('PANCANCER_SNAPTRON_JUNCTIONS_FILE') or 'tp53_junctions.txt.gz'\n",
    "\n",
    "junction_df = pd.read_table(junction_file)\n",
    "junction_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log10_mut</th>\n",
       "      <th>total_status</th>\n",
       "      <th>weight</th>\n",
       "      <th>TP53</th>\n",
       "      <th>TP53_loss</th>\n",
       "      <th>PATIENT_BARCODE</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>SUBTYPE</th>\n",
       "      <th>hypermutated</th>\n",
       "      <th>include</th>\n",
       "      <th>ID.1</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>Variant_Classification</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>1.812913</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.723395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-02-0047</td>\n",
       "      <td>GBM</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TCGA-02-0047-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wild-Type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>1.707570</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.698910</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-02-0055</td>\n",
       "      <td>GBM</td>\n",
       "      <td>IDHwt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TCGA-02-0055-01</td>\n",
       "      <td>TCGA-02-0055-01A-01D-1490-08</td>\n",
       "      <td>TP53</td>\n",
       "      <td>c.646G&gt;A</td>\n",
       "      <td>p.Val216Met</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 log10_mut  total_status    weight  TP53  TP53_loss  \\\n",
       "ID                                                                    \n",
       "TCGA-02-0047-01   1.812913           0.0  0.723395   0.0        0.0   \n",
       "TCGA-02-0055-01   1.707570           1.0  0.698910   1.0        0.0   \n",
       "\n",
       "                PATIENT_BARCODE DISEASE SUBTYPE  hypermutated  include  \\\n",
       "ID                                                                       \n",
       "TCGA-02-0047-01    TCGA-02-0047     GBM   IDHwt           0.0      1.0   \n",
       "TCGA-02-0055-01    TCGA-02-0055     GBM   IDHwt           0.0      1.0   \n",
       "\n",
       "                            ID.1          Tumor_Sample_Barcode Hugo_Symbol  \\\n",
       "ID                                                                           \n",
       "TCGA-02-0047-01  TCGA-02-0047-01                           NaN         NaN   \n",
       "TCGA-02-0055-01  TCGA-02-0055-01  TCGA-02-0055-01A-01D-1490-08        TP53   \n",
       "\n",
       "                    HGVSc        HGVSp Variant_Classification  \n",
       "ID                                                             \n",
       "TCGA-02-0047-01       NaN          NaN              Wild-Type  \n",
       "TCGA-02-0055-01  c.646G>A  p.Val216Met      Missense_Mutation  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load mutation classification scores file\n",
    "if alt_folder:\n",
    "    file = os.path.join(alt_folder,\n",
    "                        'tables', 'mutation_classification_scores.tsv')\n",
    "else:\n",
    "    file = os.path.join('..', '..', 'classifiers', 'TP53',\n",
    "                        'tables', 'mutation_classification_scores.tsv')\n",
    "\n",
    "mut_scores_df = pd.read_table(file, index_col=0)\n",
    "mut_scores_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gway/anaconda3/envs/pancancer-classifier/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>ExAC_AF_NFE</th>\n",
       "      <th>ExAC_AF_OTH</th>\n",
       "      <th>ExAC_AF_SAS</th>\n",
       "      <th>GENE_PHENO</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>COSMIC</th>\n",
       "      <th>CENTERS</th>\n",
       "      <th>CONTEXT</th>\n",
       "      <th>DBVS</th>\n",
       "      <th>NCALLERS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACC2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>10</td>\n",
       "      <td>123810032</td>\n",
       "      <td>123810032</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SITE|p.T38M|c.113C&gt;T|3</td>\n",
       "      <td>MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS</td>\n",
       "      <td>GGACACGCCCG</td>\n",
       "      <td>by1000G</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JAKMIP3</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>10</td>\n",
       "      <td>133967449</td>\n",
       "      <td>133967449</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS</td>\n",
       "      <td>CTGGACGAGGA</td>\n",
       "      <td>byFrequency</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PANX3</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>11</td>\n",
       "      <td>124489539</td>\n",
       "      <td>124489539</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SITE|p.R296Q|c.887G&gt;A|3</td>\n",
       "      <td>MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS</td>\n",
       "      <td>ATGTCGGTGGG</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>11</td>\n",
       "      <td>47380512</td>\n",
       "      <td>47380512</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>RADIA|MUSE</td>\n",
       "      <td>GGCTGGGGACA</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>11</td>\n",
       "      <td>89868837</td>\n",
       "      <td>89868837</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SITE|p.R65C|c.193C&gt;T|4</td>\n",
       "      <td>MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS</td>\n",
       "      <td>TTCTTCGGTAA</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 114 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0       TACC2               0      .     GRCh37         10       123810032   \n",
       "1     JAKMIP3               0      .     GRCh37         10       133967449   \n",
       "2       PANX3               0      .     GRCh37         11       124489539   \n",
       "3        SPI1               0      .     GRCh37         11        47380512   \n",
       "4     NAALAD2               0      .     GRCh37         11        89868837   \n",
       "\n",
       "   End_Position Strand Variant_Classification Variant_Type   ...     \\\n",
       "0     123810032      +      Missense_Mutation          SNP   ...      \n",
       "1     133967449      +                 Silent          SNP   ...      \n",
       "2     124489539      +      Missense_Mutation          SNP   ...      \n",
       "3      47380512      +      Missense_Mutation          SNP   ...      \n",
       "4      89868837      +      Missense_Mutation          SNP   ...      \n",
       "\n",
       "  ExAC_AF_NFE ExAC_AF_OTH ExAC_AF_SAS GENE_PHENO FILTER  \\\n",
       "0           .           .           .          .   PASS   \n",
       "1           .           .           .          .   PASS   \n",
       "2           .           .           .          .   PASS   \n",
       "3           .           .           .          .   PASS   \n",
       "4           .           .           .          .   PASS   \n",
       "\n",
       "                    COSMIC                                   CENTERS  \\\n",
       "0   SITE|p.T38M|c.113C>T|3  MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS   \n",
       "1                     NONE  MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS   \n",
       "2  SITE|p.R296Q|c.887G>A|3  MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS   \n",
       "3                     NONE                                RADIA|MUSE   \n",
       "4   SITE|p.R65C|c.193C>T|4  MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS   \n",
       "\n",
       "       CONTEXT         DBVS NCALLERS  \n",
       "0  GGACACGCCCG      by1000G        5  \n",
       "1  CTGGACGAGGA  byFrequency        5  \n",
       "2  ATGTCGGTGGG            .        5  \n",
       "3  GGCTGGGGACA            .        2  \n",
       "4  TTCTTCGGTAA            .        5  \n",
       "\n",
       "[5 rows x 114 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load raw mutation file\n",
    "file = os.environ.get('PANCANCER_MUT_MAF_FILE') or os.path.join('..', '..', 'data', 'raw', 'mc3.v0.2.8.PUBLIC.maf.gz')\n",
    "\n",
    "raw_mut_df = pd.read_table(file)\n",
    "raw_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5S_rRNA</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>hsa-mir-1199</th>\n",
       "      <th>hsa-mir-150</th>\n",
       "      <th>hsa-mir-3171</th>\n",
       "      <th>hsa-mir-466</th>\n",
       "      <th>hsa-mir-5195</th>\n",
       "      <th>hsa-mir-6080</th>\n",
       "      <th>hsa-mir-7162</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_BARCODE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0047-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-02-0055-01</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 20938 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 5S_rRNA  A1BG  A1CF  A2M  A2ML1  A3GALT2  A4GALT  A4GNT  \\\n",
       "SAMPLE_BARCODE                                                             \n",
       "TCGA-02-0047-01        0     0     0    0      0        0       0      0   \n",
       "TCGA-02-0055-01        0     0     0    0      0        0       0      0   \n",
       "\n",
       "                 AAAS  AACS      ...       ZYX  ZZEF1  ZZZ3  hsa-mir-1199  \\\n",
       "SAMPLE_BARCODE                   ...                                        \n",
       "TCGA-02-0047-01     0     0      ...         0      0     0             0   \n",
       "TCGA-02-0055-01     0     0      ...         0      0     0             0   \n",
       "\n",
       "                 hsa-mir-150  hsa-mir-3171  hsa-mir-466  hsa-mir-5195  \\\n",
       "SAMPLE_BARCODE                                                          \n",
       "TCGA-02-0047-01            0             0            0             0   \n",
       "TCGA-02-0055-01            0             0            0             0   \n",
       "\n",
       "                 hsa-mir-6080  hsa-mir-7162  \n",
       "SAMPLE_BARCODE                               \n",
       "TCGA-02-0047-01             0             0  \n",
       "TCGA-02-0055-01             0             0  \n",
       "\n",
       "[2 rows x 20938 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load binary mutation file\n",
    "file = os.environ.get('PANCANCER_MUT_FILE') or os.path.join('..', '..', 'data', 'pancan_mutation_freeze.tsv')\n",
    "\n",
    "mut_df = pd.read_table(file, index_col=0)\n",
    "mut_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Next, select the samples with the specific mutation of interest\n",
    "\n",
    "Also make sure that we only select samples in which this silent mutation is the _only_ _TP53_ mutation present."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31\n",
      "20\n",
      "(20, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log10_mut</th>\n",
       "      <th>total_status</th>\n",
       "      <th>weight</th>\n",
       "      <th>TP53</th>\n",
       "      <th>TP53_loss</th>\n",
       "      <th>PATIENT_BARCODE</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>SUBTYPE</th>\n",
       "      <th>hypermutated</th>\n",
       "      <th>include</th>\n",
       "      <th>ID.1</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>Variant_Classification</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-05-4415-01</th>\n",
       "      <td>2.271842</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.642171</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-05-4415</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Not_Applicable</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TCGA-05-4415-01</td>\n",
       "      <td>TCGA-05-4415-01A-22D-1855-08</td>\n",
       "      <td>TP53</td>\n",
       "      <td>c.375G&gt;T</td>\n",
       "      <td>p.%3D</td>\n",
       "      <td>Silent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-18-3417-01</th>\n",
       "      <td>2.409933</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.719642</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-18-3417</td>\n",
       "      <td>LUSC</td>\n",
       "      <td>Not_Applicable</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TCGA-18-3417-01</td>\n",
       "      <td>TCGA-18-3417-01A-01D-1441-08</td>\n",
       "      <td>TP53</td>\n",
       "      <td>c.375G&gt;T</td>\n",
       "      <td>p.%3D</td>\n",
       "      <td>Silent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 log10_mut  total_status    weight  TP53  TP53_loss  \\\n",
       "ID                                                                    \n",
       "TCGA-05-4415-01   2.271842           0.0  0.642171   0.0        0.0   \n",
       "TCGA-18-3417-01   2.409933           0.0  0.719642   0.0        0.0   \n",
       "\n",
       "                PATIENT_BARCODE DISEASE         SUBTYPE  hypermutated  \\\n",
       "ID                                                                      \n",
       "TCGA-05-4415-01    TCGA-05-4415    LUAD  Not_Applicable           0.0   \n",
       "TCGA-18-3417-01    TCGA-18-3417    LUSC  Not_Applicable           0.0   \n",
       "\n",
       "                 include             ID.1          Tumor_Sample_Barcode  \\\n",
       "ID                                                                        \n",
       "TCGA-05-4415-01      1.0  TCGA-05-4415-01  TCGA-05-4415-01A-22D-1855-08   \n",
       "TCGA-18-3417-01      1.0  TCGA-18-3417-01  TCGA-18-3417-01A-01D-1441-08   \n",
       "\n",
       "                Hugo_Symbol     HGVSc  HGVSp Variant_Classification  \n",
       "ID                                                                   \n",
       "TCGA-05-4415-01        TP53  c.375G>T  p.%3D                 Silent  \n",
       "TCGA-18-3417-01        TP53  c.375G>T  p.%3D                 Silent  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset mutation file to samples with c375GT TP53 mutations\n",
    "silent_mut_df = (\n",
    "    raw_mut_df.query('Hugo_Symbol == \"TP53\"')\n",
    "    .query('HGVSc == \"c.375G>T\"')\n",
    ")\n",
    "\n",
    "# Obtain the samples with the specific mutation\n",
    "silent_mut_samples = silent_mut_df.Tumor_Sample_Barcode.str.slice(start=0, stop=15)\n",
    "print(len(silent_mut_samples))\n",
    "\n",
    "# From these, remove samples that also have a different mutation in TP53\n",
    "only_silent_mut_samples = (\n",
    "    mut_df.reindex(silent_mut_samples)\n",
    "    .loc[:, 'TP53']\n",
    ")\n",
    "\n",
    "only_silent_mut_samples = (\n",
    "    only_silent_mut_samples.loc[only_silent_mut_samples == 0]\n",
    "    .index\n",
    "    .tolist()\n",
    ")\n",
    "\n",
    "print(len(only_silent_mut_samples))\n",
    "\n",
    "# Select those samples in which we have mutation classification scores\n",
    "mut_silent_scores_df = (\n",
    "    mut_scores_df\n",
    "    .loc[mut_scores_df.index.isin(only_silent_mut_samples), :]\n",
    ")\n",
    "\n",
    "print(mut_silent_scores_df.shape)\n",
    "mut_silent_scores_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finally, process the junctions file and output the results for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snaptron_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>left_motif</th>\n",
       "      <th>right_motif</th>\n",
       "      <th>samples_count</th>\n",
       "      <th>coverage_sum</th>\n",
       "      <th>coverage_avg</th>\n",
       "      <th>...</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>SUBTYPE</th>\n",
       "      <th>hypermutated</th>\n",
       "      <th>include</th>\n",
       "      <th>ID.1</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>Variant_Classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13945731</td>\n",
       "      <td>7676009</td>\n",
       "      <td>7676046</td>\n",
       "      <td>38</td>\n",
       "      <td>-</td>\n",
       "      <td>GT</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>OV</td>\n",
       "      <td>Not_Applicable</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-25-2404-01</td>\n",
       "      <td>TCGA-25-2404-01A-01W-0799-08</td>\n",
       "      <td>TP53</td>\n",
       "      <td>c.641A&gt;G</td>\n",
       "      <td>p.His214Arg</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>13945751</td>\n",
       "      <td>7676064</td>\n",
       "      <td>7676130</td>\n",
       "      <td>67</td>\n",
       "      <td>+</td>\n",
       "      <td>GT</td>\n",
       "      <td>AG</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>OV</td>\n",
       "      <td>Not_Applicable</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-25-2404-01</td>\n",
       "      <td>TCGA-25-2404-01A-01W-0799-08</td>\n",
       "      <td>TP53</td>\n",
       "      <td>c.641A&gt;G</td>\n",
       "      <td>p.His214Arg</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    snaptron_id    start      end  length strand left_motif right_motif  \\\n",
       "13     13945731  7676009  7676046      38      -         GT          AG   \n",
       "14     13945751  7676064  7676130      67      +         GT          AG   \n",
       "\n",
       "    samples_count  coverage_sum  coverage_avg           ...            \\\n",
       "13              1             1           1.0           ...             \n",
       "14              6             6           1.0           ...             \n",
       "\n",
       "    DISEASE         SUBTYPE  hypermutated  include             ID.1  \\\n",
       "13       OV  Not_Applicable           0.0      0.0  TCGA-25-2404-01   \n",
       "14       OV  Not_Applicable           0.0      0.0  TCGA-25-2404-01   \n",
       "\n",
       "            Tumor_Sample_Barcode Hugo_Symbol     HGVSc        HGVSp  \\\n",
       "13  TCGA-25-2404-01A-01W-0799-08        TP53  c.641A>G  p.His214Arg   \n",
       "14  TCGA-25-2404-01A-01W-0799-08        TP53  c.641A>G  p.His214Arg   \n",
       "\n",
       "    Variant_Classification  \n",
       "13       Missense_Mutation  \n",
       "14       Missense_Mutation  \n",
       "\n",
       "[2 rows x 34 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Process and output junction file\n",
    "out_file = os.environ.get('PANCANCER_JUNCTIONS_MUTIONS_OUTPUT_FILE') or 'tp53_junctions_with_mutations.csv.gz'\n",
    "\n",
    "junctions_process_df = (\n",
    "    junction_df.assign(diff_start = abs(7675994 - junction_df['start']),\n",
    "                       diff_end = abs(7675994 - junction_df['end']))\n",
    "    .sort_values(by = \"diff_start\")\n",
    ")\n",
    "\n",
    "junctions_process_df = (\n",
    "    junctions_process_df\n",
    "    .assign(rail_id = junctions_process_df.apply(get_rail_id, axis=1))\n",
    "    .set_index(['snaptron_id', 'start', 'end', 'length', 'strand',\n",
    "                'left_motif', 'right_motif', 'samples_count',\n",
    "                'coverage_sum', 'coverage_avg', 'coverage_median',\n",
    "                'diff_start', 'diff_end'])['rail_id']\n",
    "    .apply(pd.Series)\n",
    "    .stack()\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "junctions_process_df[0] = junctions_process_df[0].astype(int)\n",
    "\n",
    "junctions_process_df = (\n",
    "    junctions_process_df\n",
    "    .merge(dictionary_df, left_on=0, right_on='rail_id')\n",
    ")\n",
    "\n",
    "junctions_process_df = (\n",
    "    junctions_process_df\n",
    "    .assign(\n",
    "        tcga_id = junctions_process_df['gdc_cases.samples.submitter_id']\n",
    "        .str\n",
    "        .slice(start=0, stop=15)\n",
    "    )\n",
    "    .merge(mut_scores_df, left_on='tcga_id', right_index=True)\n",
    ")\n",
    "\n",
    "# FIXME, compression to infer, on pandas update 0.24.0+\n",
    "if out_file.endswith('.gz'):\n",
    "    compression_type = 'gzip'\n",
    "else:\n",
    "    compression_type = None\n",
    "junctions_process_df.to_csv(out_file, compression=compression_type)\n",
    "junctions_process_df.head(2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:pancancer-classifier]",
   "language": "python",
   "name": "conda-env-pancancer-classifier-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
